### PR TITLE
Delete libspdm_verify_cert_signature_algo_OID

### PR DESCRIPTION
**Pull request by Wenxing-hou**
Fix: #1093

The leaf cert signature algo is different from libspdm negotiated base_hash_algo/aysm_algo. Because the leaf cert signature algo is used from the previous cert.
For a right leaf cert, the Libspdm negotiated bash_aysm_algo and cert_pubkey_algo should be consistent.

There are four patch to optimize and fix `libspdm_x509_certificate_check` :

**1. Add check for the consistence of Libspdm negotiated bash_aysm_algo and cert_pubkey_algo.
2. Add negotiated algo and cert public key algo mismatched case for unit test.
3. Add some cert from big website to test.
4. Make libspdm_x509_certificate_check clear by using existing function.**

Signed-off-by: Wenxing Hou <wenxing.hou@intel.com>

### CODE REVIEW

File: library/spdm_crypt_lib/libspdm_crypt_crypt.c [hunk 1]
```diff
@@ -4382,10 +4162,11 @@ bool libspdm_x509_certificate_check(const uint8_t *cert, size_t cert_size,
         goto cleanup;
     }
 
-    /* 3. verify sinature_algorithem*/
-    status =
-        libspdm_verify_cert_signature_algo_OID(cert, cert_size, base_asym_algo, base_hash_algo);
-    if (!status) {
+    /* 3. sinature_algorithem*/
+    value = 0;
+    status = libspdm_x509_get_signature_algorithm(cert, cert_size, NULL, &value);
```
Comment (jyao1) [inline library/spdm_crypt_lib/libspdm_crypt_crypt.c:4167]: why we need this?

Comment (Wenxing-hou) [reply library/spdm_crypt_lib/libspdm_crypt_crypt.c]: I have changed the code. Thanks.

File: include/hal/library/cryptlib.h [hunk 2]
```diff
@@ -1740,6 +1740,25 @@ bool libspdm_x509_get_issuer_name(const uint8_t *cert, size_t cert_size,
                                   uint8_t *cert_issuer,
                                   size_t *issuer_size);
 
+/**
+ * Retrieve the public key encrypt algo oid from one X.509 certificate.
+ *
+ * @param[in]      cert                 Pointer to the DER-encoded X509 certificate.
+ * @param[in]      cert_size            size of the X509 certificate in bytes.
+ * @param[in,out]  cert_public_key_raw  the public key raw data in cert, only rsa public key will output
```
Comment (steven-bellock) [inline include/hal/library/cryptlib.h:1748]: This is only for RSA public keys?

File: library/spdm_crypt_lib/libspdm_crypt_crypt.c [hunk 3]
```diff
@@ -4476,7 +4476,7 @@ bool libspdm_x509_certificate_check(const uint8_t *cert, size_t cert_size,
 
     /* 3. verify sinature_algorithem*/
```
Comment (steven-bellock) [inline library/spdm_crypt_lib/libspdm_crypt_crypt.c:4477]: /* 3. Verify public key algorithm. */

File: os_stub/cryptlib_openssl/pk/x509.c [hunk 4]
```diff
@@ -28,6 +28,68 @@
...
+static const uint8_t m_libspdm_oid_key_encry_rsa4096_algo[] = KEY_ENCRY_ALGO_RSA4096_FLAG;
+
+static const uint8_t m_libspdm_oid_key_encry_rsa2048_algo_other[] =
+    KEY_ENCRY_ALGO_RSA2048_FLAG_OTHER;
+static const uint8_t m_libspdm_oid_key_encry_rsa3072_algo_other[] =
+    KEY_ENCRY_ALGO_RSA3072_FLAG_OTHER;
+static const uint8_t m_libspdm_oid_key_encry_rsa4096_algo_other[] =
+    KEY_ENCRY_ALGO_RSA4096_FLAG_OTHER;
+
+
```
Comment (steven-bellock) [inline os_stub/cryptlib_openssl/pk/x509.c:85]: Remove line.

File: os_stub/cryptlib_null/pk/x509.c [hunk 5]
```diff
@@ -462,6 +462,29 @@ bool libspdm_x509_get_issuer_name(const uint8_t *cert, size_t cert_size,
...
+bool libspdm_x509_get_public_key_info(const uint8_t *cert, size_t cert_size,
+                                      uint8_t *cert_public_key_raw,
+                                      size_t *public_key_raw_size,
+                                      uint8_t *oid, size_t *oid_size)
+{
+    LIBSPDM_ASSERT(false);
+    return false;
+}
+
+
```
Comment (steven-bellock) [inline os_stub/cryptlib_null/pk/x509.c:487]: Remove line.

**Review by steven-bellock — CHANGES_REQUESTED** [review]

File: include/hal/library/cryptlib.h (hunk 2 above)
Comment (Wenxing-hou) [reply include/hal/library/cryptlib.h]: Thanks. I have changed the comment. This is only for RSA and ECC key. 
In OpenSSL, the `public_key_size = i2d_PublicKey(public_key, NULL);` will be failed for ED448/ED25519.
In MbedTLS, it only support RSA and ECC now.

**Review by steven-bellock — APPROVED** [review]
Should follow up with a unit test for the Requester where the public key algorithm of the leaf certificate returned by the Responder does not match the negotiated algorithm.

File: include/library/spdm_crypt_lib.h [hunk 6]
```diff
@@ -1258,15 +1258,14 @@ bool libspdm_get_random_number(size_t size, uint8_t *rand);
...
- * @param[in]  base_hash_algo        SPDM base_hash_algo
  * @param[in]  is_device_cert_model  If true, the cert chain is DeviceCert model;
  *                                   If false, the cert chain is AliasCert model;
  *
  * @retval  true   Success.
  * @retval  false  Certificate is not valid
  **/
 bool libspdm_x509_certificate_check(const uint8_t *cert, size_t cert_size,
-                                    uint32_t base_asym_algo, uint32_t base_hash_algo,
+                                    uint32_t base_asym_algo,
```
Comment (jyao1) [inline include/library/spdm_crypt_lib.h:1268]: Please dont change public API.

File: include/library/spdm_crypt_lib.h [hunk 7]
```diff
@@ -1344,8 +1342,7 @@ libspdm_get_dmtf_subject_alt_name(const uint8_t *cert, const size_t cert_size,
  * @retval false certificate chain data integrity verification fail.
  **/
 bool libspdm_verify_cert_chain_data(uint8_t *cert_chain_data, size_t cert_chain_data_size,
-                                    uint32_t base_asym_algo, uint32_t base_hash_algo,
-                                    bool is_device_cert_model);
+                                    uint32_t base_asym_algo, bool is_device_cert_model);
```
Comment (jyao1) [inline include/library/spdm_crypt_lib.h:1345]: Please dont change public API.

File: unit_test/sample_key/test_web_cert/readme.txt [hunk 8]
```diff
@@ -0,0 +1 @@
+Some cert from big website to test.
```
Comment (jyao1) [inline unit_test/sample_key/test_web_cert/readme.txt:1]: Please describe clearly where you get these.

File: include/hal/library/cryptlib.h [hunk 9]
```diff
@@ -1740,6 +1740,25 @@ bool libspdm_x509_get_issuer_name(const uint8_t *cert, size_t cert_size,
...
+ * @param[in,out]  cert_public_key_raw  the public key raw data in cert, only rsa and ecc public key will output
+ * @param[in,out]  public_key_raw_size  The size of the public key raw data in cert
+ * @param[out]     oid                  public key encrypt algo oid
+ * @param[out]     oid_size             the size of public key encrypt algo oid
+ *
+ * @retval  true   Retrieve the public key encrypt algo oid successfully.
+ * @retval  false  Invalid certificate, or the oid_size/public_key_raw_size is too small for the result.
+ **/
+
+bool libspdm_x509_get_public_key_info(const uint8_t *cert, size_t cert_size,
```
Comment (jyao1) [inline include/hal/library/cryptlib.h:1757]: I am wondering why we need add a new API.

Can we just reuse existing APIs to get the public_key_info in spdm_crypt_lib?

Comment (Wenxing-hou) [reply include/hal/library/cryptlib.h]: I have changed the code and deleted the new API. Thanks.

**Pull request merged**